Allow clustermgtd to handle compute nodes rebooted via `scontrol reboot`

jdeamicis opened the pull request:
### Description of changes
* Fix logic of clustermgtd to deal with EC2 status checks failures when rebooting compute nodes with `scontrol reboot`:
  * Do not consider nodes failing EC2 health checks as unhealthy while they are in Slurm state rebooting (`REBOOT_ISSUED`).
  * Do not consider nodes failing EC2 health checks as unhealthy for a certain time after Slurmd has started (EC2 may be late in updating the instance status checks).
  * Relates to https://github.com/aws/aws-parallelcluster/issues/4751

### Tests
* Added unit test scenarios to validate new logic.
* Added integration test to validate new logic:

### References
* https://github.com/aws/aws-parallelcluster/pull/4838

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Fix small bug in slurm_commands.py

Fix small bug that could cause an inconsistent stripping of spaces
across the fields extracted from scontrol show nodes

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add slurmdstarttime attribute to SlurmNode class

slurmdstarttime will track the SlurmdStartTime field returned by scontrol
show nodes, which is the instant at which the Slurmd daemon started on
the compute node.

Add some typing hints in the code when slurmdstarttime is defined and used.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Do not drain nodes failing EC2 checks after scontrol reboot

Add a new timeout that controls how long after the booting of slurmd
(which happens at the first start of a compute node or after
an scontrol reboot operation) clustermgtd is allowed to drain a compute
node if the corresponding EC2 instance has the EC2 status checks failing.

After performing scontrol reboot, EC2 might signal the instance as unhealthy
for a certain amount of time, even after the instance has fully booted and
the Slurm compute node has registered itself to the Slurm control daemon. This
is just an issue of timing in updating the status of the instance on the EC2
side.

This is related to https://github.com/aws/aws-parallelcluster/issues/4751

Add some logging about nodes failing EC2 health checks that were not set
to DRAIN because they had their Slurmd daemon recently restarted.

Add unit test case for time_is_up() to cover case when initial_time is None.

Add python typing hint to time_is_up()

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Do not drain compute nodes for EC2 checks while rebooting

Set default timezone to UTC when reading SlurmdStartTime.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Fix scontrol show nodes parser

Add grep lookbehind logic to exclude "NextState=" when searching
for "State=".

When rebooting nodes via scontrol reboot, scontrol show node shows
a field "NextState=". The previous parsing syntax caught that as
a second instance of "State=".

Add unit test for the scontrol show node output parser.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Fix code-linters failures

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Adapt test_handle_health_check for scontrol reboot scenarios

Add set of scenarios to check the new logic of handling nodes that fail
the EC2 health checks due to an scontrol reboot.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Fix issues after rebasing onto upstream/develop

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add entry to CHANGELOG.md

Add entry to the CHANGELOG about the new logic of clustermgtd when
handling EC2 status checks failures for nodes rebooted via Slurm.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

chenwany commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -324,7 +329,7 @@ def _get_partition_nodes(partition_name, command_timeout=DEFAULT_GET_INFO_COMMAN
 
 
-def _parse_nodes_info(slurm_node_info):
+def _parse_nodes_info(slurm_node_info: str) -> List[SlurmNode]:
```
Could you update the example below with `SlurmdStartTime` and `NextState`?

jdeamicis replied:
Done.

chenwany commented on src/slurm_plugin/slurm_resources.py:
```diff
@@ -82,6 +82,7 @@ def __eq__(self, other):
 
 
 class SlurmNode(metaclass=ABCMeta):
+    slurmdstarttime: datetime
```
Instead of leaving it here, what about we change line 113 to be `def __init__(...., slurmdstarttime: datetime = None)` ?

jdeamicis replied:
Done.

jdeamicis replied:
Done.

chenwany commented on tests/common/schedulers/test_slurm_commands.py:
```diff
@@ -83,82 +86,122 @@ def test_is_static_node(nodename, expected_is_static):
+            "NodeHostName=multiple-dy-c5xlarge-4\n"
+            "State=IDLE+CLOUD+POWER\n"
+            "Partitions=multiple,multiple2\n"
+            "Reason=(Code:InsufficientInstanceCapacity)Failure when resuming nodes \n"
```
With the changes in code: `(Reason=.+) |(######)` to `(Reason=.*)|(######)` Looks like we are passing more information to Reasons, could you change this test case accordingly?

jdeamicis replied:
What do you mean here exactly?

chenwany replied:
with `scontrol show nodes` result:
`Reason=test [root@2023-01-31T21:24:55]`
Previously, we are using regex `(Reason=.+) |(######)`, with that expression, the match will be `Reason=test `
With your change, `(Reason=.*)|(######)`, the match will be `Reason=test [root@2023-01-31T21:24:55]`
Can we include that in this test case?

jdeamicis replied:
Done.

chenwany commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -362,15 +367,21 @@ def _parse_nodes_info(slurm_node_info):
         for line in lines:
             key, value = line.split("=")
+            if key == "SlurmdStartTime":
+                if value != "None":
```
minor: line 381- 384 can be simplify into one line: `value = datetime.strptime(value, "%Y-%m-%dT%H:%M:%S").astimezone(tz=timezone.utc) if value != "None" else None`

jdeamicis replied:
Unfortunately the line becomes too long and the autoformatter splits it into even more lines that it's using now. I'm going to keep it as it is.

chenwany commented on src/common/utils.py:
```diff
@@ -250,11 +250,13 @@ def convert_range_to_list(node_range):
 
 
-def time_is_up(initial_time, current_time, grace_time):
+def time_is_up(initial_time: datetime, current_time: datetime, grace_time: int):
```
Is it true that restrict grace_time is always int?
In `insufficient_capacity_timeout`, we allows input as float:
        self.insufficient_capacity_timeout = config.getfloat(
            "clustermgtd", "insufficient_capacity_timeout", fallback=self.DEFAULTS.get("insufficient_capacity_timeout")
        )

jdeamicis replied:
Changing it to `float`, which in Python typing should cover also the cases when an integer is passed (see [The numeric tower](https://peps.python.org/pep-0484/#the-numeric-tower)).

jdeamicis replied:
Done.

chenwany commented on src/slurm_plugin/clustermgtd.py:
```diff
@@ -956,17 +965,30 @@ def _handle_protected_mode_process(self, active_nodes, partitions_name_map):
+                set_nodes_drain(nodes_name_failing_health_check, reason=f"Node failing {health_check_type}")
+            if len(nodes_name_recently_rebooted) > 0:
+                log.info(
+                    "These nodes were recently restarted and they were not set to DRAIN: %s",
```
What about changing `restarted` to rebooted?

jdeamicis replied:
Minor discussion here: strictly speaking, this condition may occur even if someone manually restarts `slurmd` on a compute node, without completely rebooting the compute node. I believe I had originally written something like `On these nodes the slurmd daemon was recently restarted...`, and that's why I kept the restarted word here.

Also, this condition may also occur the first time ever a compute node is booted, so _rebooted_ or _restarted_ may be the wrong words. 

Should we write something like: `The slurmd service was recently started on these nodes and therefore they were not set to DRAIN`? Do we like it or is it too convoluted?

jdeamicis committed:
Implement first round of code reviews for the PR

Use single quotes instead of double quotes in grep part of
SCONTROL_OUTPUT_AWK_PARSER to protect from possible bash
interpretation of the ! character.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add comments in _handle_nodes_failing_health_check

Improve unit test for SCONTROL_OUTPUT_AWK_PARSER

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Adapt reasons in unit tests to format of SCONTROL_OUTPUT_AWK_PARSER

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

demartinofra commented on src/common/schedulers/slurm_commands.py:
```diff
@@ -362,15 +371,21 @@ def _parse_nodes_info(slurm_node_info):
+    node_info = slurm_node_info.split("######\n")
     slurm_nodes = []
     for node in node_info:
-        lines = node.strip().splitlines()
```
why did you remove the strip()? it could be useful in case for any reason there are trailing whitespaces in the Slurm output

jdeamicis replied:
I just checked. Actually this has more to do with the output of our `SCONTROL_OUTPUT_AWK_PARSER` than anything coming from `scontrol show node`: this `strip()` was removing the trailing `\n` at the end each node line, which was there because we were splitting by `######` and not `######\n`.

Now, since we split according to `######\n`, it's not strictly necessary. Actually, the `strip()` in that position would remove trailing spaces **only** in the last field we parse in our awk parser, which currently is `Reason`. If in the future we start parsing additional fields (`Comment` or `Extra`, for example), the `strip()` will start affecting a field different than the reason.

Given this information, I would be more inclined towards not having the `strip()` at all.

demartinofra commented on src/slurm_plugin/clustermgtd.py:
let's slightly change this message to make it more explicit: Ignoring health check failure due to reboot for nodes:

jdeamicis replied:
Done.

jdeamicis committed:
Address comment in CR about logging message

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
